quinn-proto/src/connection/assembler.rs

```diff
@@ -106,35 +118,55 @@ impl Assembler {
         let new = BinaryHeap::with_capacity(self.data.len());
         let old = mem::replace(&mut self.data, new);
         for chunk in old.into_sorted_vec().into_iter().rev() {
-            let end = offset + (buffer.len() as u64);
-            if let Some(overlap) = end.checked_sub(chunk.offset) {
-                if let Some(bytes) = chunk.bytes.get(overlap as usize..) {
-                    buffer.extend_from_slice(bytes);
+            if chunk.should_defragment(include_highly_utilized) {
+                let end = offset + (buffer.len() as u64);
+                if let Some(overlap) = end.checked_sub(chunk.offset) {
+                    if let Some(bytes) = chunk.bytes.get(overlap as usize..) {
+                        buffer.extend_from_slice(bytes);
+                    }
+                } else {
+                    let bytes = buffer.split().freeze();
+                    self.data.push(Buffer {
+                        offset,
+                        bytes,
+                        over_allocation: 0,
+                    });
+                    offset = chunk.offset;
+                    buffer.extend_from_slice(&chunk.bytes);
                 }
             } else {
-                let bytes = buffer.split().freeze();
-                self.data.push(Buffer { offset, bytes });
-                offset = chunk.offset;
-                buffer.extend_from_slice(&chunk.bytes);
+                self.data.push(chunk);
             }
         }
 
         let bytes = buffer.split().freeze();
-        self.data.push(Buffer { offset, bytes });
-        self.defragmented = self.data.len();
-    }
-
-    pub(crate) fn insert(&mut self, mut offset: u64, mut bytes: Bytes) {
+        self.data.push(Buffer {
+            offset,
+            bytes,
+            over_allocation: 0,
+        });
+        self.over_allocation = if include_highly_utilized {
+            0
+        } else {
+            high_utilization_over_allocation
+        };
+    }
+
+    pub(crate) fn insert(&mut self, mut offset: u64, mut bytes: Bytes, mut allocation_size: usize) {
         self.end = self.end.max(offset + bytes.len() as u64);
-
         if let State::Unordered { ref mut recvd } = self.state {
             // Discard duplicate data
             for duplicate in recvd.replace(offset..offset + bytes.len() as u64) {
                 if duplicate.start > offset {
+                    let over_allocation = (duplicate.end - duplicate.start) as usize;
                     self.data.push(Buffer {
                         offset,
                         bytes: bytes.split_to((duplicate.start - offset) as usize),
+                        over_allocation,
                     });
+                    self.over_allocation += over_allocation;
+                    allocation_size =
+                        allocation_size.saturating_sub((duplicate.end - offset) as usize);
```

Comment:
Our main concern here is ensuring that there is no possible pattern of insert and read calls that can cause us to be keeping substantially more memory allocated than the amount of useful data we're actually storing, which is the figure governed by flow control. I think this pattern fails that test due to allowing O(n^2) memory use for n useful bytes: consider a flow control limit of `n` bytes, and an ordered sequence of inserts covering ranges `n-1..n, n-2..n, n-3..n`, ...

Treating each chunk as if it had a completely independent allocation does risk counting the same allocation multiple times, but I think risking unnecessary defragmentation passes is preferable to risking quadratic memory overhead.

Reply:
The new version now treats each chunk as a completely independent allocation.